backend/routes/nested-path-example/orm.js

```diff
@@ -1,28 +1,50 @@
 import DbRow from "../../db/models/DbRow.js";
 import { Router } from "express";
 import asyncHandler from "express-async-handler";
+import checkJwt from "../../middleware/checkJwt.js";
 
 const router = Router();
 
+const routeHandler = asyncHandler(async (req, res) => {
+  const { rowId } = req.params;
+
+  const allRows = await DbRow.findAll();
+  const specificRow = await DbRow.findByPk(rowId);
+
+  if (!specificRow) {
+    throw new Error(`Requested row does not exist: ${rowId}`);
+  }
+
+  res.json({ rowId, specificRow, numRows: allRows.length });
+});
+
+// example with Sequelize ORM 🔒 protected by Keycloak
+// http://0.0.0.0:8100/nested-path-example/orm-2
+router.get(
+  "/orm-2",
+
+  // protect this route with JWT authentication
+  checkJwt,
```

Comment:
I'm sure there's some magic going on here, but I could proably use a walk-through of how this actually works.

Reply:
Definitely, I can go over it in more detail but the short version is that basically everything in Express is "middleware" and it all gets processed in order for each request. The "router" definitions are also middleware technically. So the `checkJwt` function basically runs before the actual URL handler function, and it verifies your token before anything else is allowed to happen.

So the order of middleware here is like:
- CORs, logging, other stuff defined in index.js
- Route handler stuff (see below)
- Finally there's an error handler, also in index.js, if the Route handler failed

The signature for `express.use` is like:
(`url match pattern, middleware function[, middleware function, ...]`)

For a route definition, the last function in the chain is the actual handler that returns to the browser, but anything you put before that gets run first. So the `checkJwt` function checks your access token and cancels it with a 401 error if the token is invalid.

Reply:
...and on that note, I think we'll use a similar middleware pattern to check the specific roles

1. token valid?
2. can edit data?
3. can edit THIS data?
4. if so, then run the handler function